runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -782,10 +783,23 @@ function M.get_virtual_text_chunks_for_line(bufnr, line, line_diags, opts)
   -- TODO(tjdevries): Allow different servers to be shown first somehow?
   -- TODO(tjdevries): Display server name associated with these?
   if last.message then
+    local stripped_msg = last.message:gsub("\r", "")
+    local newline_index, _ = string.find(stripped_msg, "\n")
+    local diagnostic_msg
+    if newline_index then
+      if opts.truncated then
+        diagnostic_msg = string.sub(stripped_msg, 1, newline_index - 1)
+      else
+        diagnostic_msg = stripped_msg:gsub("\n", " ")
+      end
+    else
+      diagnostic_msg = stripped_msg
+    end
+
```

Comment:
maybe a better solution would be to expose a option that is a function that takes in the `function(client, message) -> message`. This way people can do whatever they want with the message and we don't have a multiplication of all the possible configurations someone could have.

What do you think about doing that?

(Also, I'd be fine with changing our  default to only show the first line and a `...` if there is more text on other lines or something, since otherwise the virtual text can be way too big).

Reply:
> maybe a better solution would be to expose a option that is a function that takes in the function(client, message) -> message. This way people can do whatever they want with the message and we don't have a multiplication of all the possible configurations someone could have.

In some way I can see the benefit of that for sure, but the negative side of this would be that there still isn't an easy way to have a sane default. It's either, you get everything (even though they may be insanely long), or you write your own function to deal with it. The reason I liked this approach was it was just a simple `truncated = true` that would be required.

I'd be way more inclined to do your second suggestion:

> (Also, I'd be fine with changing our default to only show the first line and a ... if there is more text on other lines or something, since otherwise the virtual text can be way too big).

Since I think this is way more sane. If you're ok with doing this, I'm ok with making the change.

Reply:
Ah, sorry, I was saying we could do _both_ of those. The default `function(client, message) -> message` would return the truncated string. Or perhaps it would be better to pass as a table, so we can change the arguments later. `{ client = client, message = message, ... }` might be easier to extend later.

I think we can do it both ways (and this would increase the configurability as time goes on).

Reply:
(the function + changing the default means we don't have to provide `truncated` as a new configuration option though)

Reply:
ah ok, gotcha gotcha. Sounds good. I'll take a stab at this then and offer up both 👍🏼

Reply:
Quick question on this. Just to make sure I'm understanding, this function would still be exposed as an option under `virtual_text` though right? So if someone was going to use it, it'd be 

```lua
local my_function_to_handle_diagnotsic_virtual_text = function({client, message}) ... end

vim.lsp.with(vim.lsp.diagnostic.on_publish_diagnostics, {
  virtual_text = {
    display = my_function_to_handle_diagnotsic_virtual_text -- is display a good name?
  }
})
```
I just want to make sure I'm following how you envision this.

Reply:
Yeah, exactly that style. 

I'd say something like `get_message` would be better -- something with a verb gives a hint that it's a function.